application/cms/models.py

```diff
@@ -102,18 +102,83 @@ def process(value):
 
 
 class FrequencyOfRelease(db.Model):
+    __tablename__ = "frequency_of_release"
+
     id = db.Column(db.Integer, primary_key=True)
     description = db.Column(db.String(), nullable=False)
     position = db.Column(db.Integer, nullable=False)
 
 
 class TypeOfStatistic(db.Model):
+    __tablename__ = "type_of_statistic"
+
     id = db.Column(db.Integer, primary_key=True)
     internal = db.Column(db.String(), nullable=False)
     external = db.Column(db.String(), nullable=False)
     position = db.Column(db.Integer, nullable=False)
 
 
+class DataSource(db.Model):
+    __tablename__ = "data_source"
+
+    # columns
+    id = db.Column(db.Integer, primary_key=True)
+    title = db.Column(db.String(255), nullable=True)
+
+    type_of_data = db.Column(ArrayOfEnum(db.Enum(TypeOfData, name="type_of_data_types")), default=[], nullable=True)
+    type_of_statistic_id = db.Column(db.Integer, nullable=True)
+
+    publisher_id = db.Column(db.String(255), nullable=True)
+    external_url = db.Column(db.String(255), nullable=True)
+    publication_date = db.Column(db.String(255), nullable=True)
+    note_on_corrections_or_updates = db.Column(db.TEXT, nullable=True)
+
+    frequency_of_release_id = db.Column(db.Integer, nullable=True)
+    frequency_of_release_other = db.Column(db.TEXT, nullable=True)
+
+    purpose = db.Column(db.TEXT, nullable=True)
+
+    # relationships
+    type_of_statistic = relationship("TypeOfStatistic", foreign_keys=[type_of_statistic_id])
+    publisher = relationship("Organisation", foreign_keys=[publisher_id], backref="data_sources")
+    frequency_of_release = relationship("FrequencyOfRelease", foreign_keys=[frequency_of_release_id])
+
+    __table_args__ = (
+        ForeignKeyConstraint(["type_of_statistic_id"], ["type_of_statistic.id"]),
+        ForeignKeyConstraint(["publisher_id"], ["organisation.id"]),
+        ForeignKeyConstraint(["frequency_of_release_id"], ["frequency_of_release.id"]),
+    )
+
+    def to_dict(self):
+        return {
+            "id": self.id,
+            "title": self.title,
+            "type_of_data": self.type_of_data,
+            "type_of_statistic": self.type_of_statistic,
+            "publisher": self.publisher,
+            "external_url": self.external_url,
+            "publication_date": self.publication_date,
+            "note_on_corrections_or_updates": self.note_on_corrections_or_updates,
+            "frequency_of_release": (
+                self.frequency_of_release if self.frequency_of_release else self.frequency_of_release_other
+            ),
+            "purpose": self.purpose,
+        }
+
+
+class DataSourceInPage(db.Model):
```

Comment:
`DataSourceForPage`? Do we also need some kind of mechanism here for distinguishing primary and secondary sources?

Reply:
Maybe we want some kind of explicit `position` value (1 for primary, 2 for secondary, etc)? Good question.

Reply:
@samuelhwilliams hard to know for sure, but I think that primary and secondary sources are semantically a bit different from position? (eg we used to let people have 2 secondary sources of equal weight).

That said, perhaps "primary" and "secondary" are the wrong descriptions - my vague memory are that "secondary sources" are sources which in turn are referencing primary sources? Whereas in our purposes both are usually "primary sources" (eg the data source for the subject and the data source for the population) but one is used as the main measure (aka numerator) and the other as the comparison (aka denominator).

Words are hard... 🤷‍♂️

Reply:
Yeah, my understanding of our use of 'primary' and 'secondary' is that it's simply 'first' and 'second' source. It's definitely not at all clear if we mean the context-appropriate definition of 'primary' and 'secondary' (i.e. first-hand/second-hand information). In our use, it seems fine to have a 'position' thing if we think we need that.

It feels like something we might need in the future (when we start to de-duplicate sources). We can maybe get away with not having it now by ordering on data source ID.

Reply:
@samuelhwilliams ok, let's postpone figuring out how to order these explicitly for now, and stick with using the ID in the short term.